File: demo2/glibrary/src/main/java/com/gangtise/gdk/result/GDataSet.java
package com.gangtise.gdk.result;

import android.os.Build;
import android.util.Log;

import androidx.annotation.RequiresApi;

import com.alibaba.fastjson.JSONObject;
import com.gangtise.gdk.common.Constant;
import com.gangtise.gdk.common.ContentMap;
import com.gangtise.gdk.common.GtsArrayList;
import com.gangtise.gdk.result.dataprocess.GColumnInfo;
import com.gangtise.gdk.result.dataprocess.GDataTool;
import com.gangtise.gdk.result.dataprocess.Table;
import com.gangtise.gdk.result.dataprocess.TableCellData;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class GDataSet implements Serializable {

    public String id; //请求对象id
    public String name;//表名
    public Integer begin;//开始位置
    public Integer rowCount = 0;//行数
    public Integer colCount = 0; //列数
    public Integer totalRows; //总数
    public List<GColumnInfo> columnInfos;  //表头信息

    //table -> data
    private List<List<Object>> data;
    //table -> flags
    private JSONObject flags;
    //col 列模式、row行模式
    public String rank;

    public void setData(List<List<Object>> data) {
        this.data = data;
    }

    public void setFlags(JSONObject flags) {
        this.flags = flags;
    }


    /**
     * 根据序号获取列表信息的对象
     * @param {int} colIndex 列号
     * @returns {GColumnInfo} 表信息
     */
    @RequiresApi(api = Build.VERSION_CODES.N)
    public GColumnInfo getColumn(int colIndex){
        List<GColumnInfo> list = columnInfos.stream().filter(t -> t.index==colIndex).collect(Collectors.toList());

        GColumnInfo result = new GColumnInfo();
        if(list.size()>0){
            result = list.get(0);
        }
        return result;
    }

    /**
     * 根据名称获取列表信息的对象
     * @param colName  列字段名
     * @return 表信息
     */
    @RequiresApi(api = Build.VERSION_CODES.N)
    public GColumnInfo getColumnByName(String colName){
        List<GColumnInfo> list = columnInfos.stream().filter(t -> t.name.toLowerCase().equals(colName.toLowerCase()))
                .collect(Collectors.toList());

        GColumnInfo result = new GColumnInfo();
        if(list.size()>0){
            result = list.get(0);
        }
        return result;
    }



    /**
     * 根据行号，列号获取单元格对象
     * @param {int} rowIndex  行号
     * @param {int} colIndex  列号
     * @return {GCellData} cellData 单元格数据
     *
     */
    @RequiresApi(api = Build.VERSION_CODES.N)
    public TableCellData getCellData(int rowIndex, int colIndex){
        int rIndex = rowIndex -1;
        int cIndex = colIndex -1;

        Table table = ContentMap.SOURCE_TABLE_MAP.get(id);
        if(table==null){
            Log.e("GDataSet","ContentMap.SOURCE_TABLE_MAP对应的请求id数据为空！");
            return null;
        }
        List<List<Object>> data = table.getData();
        if(data.size()==0){
            Log.e("GDataSet","table.getData()对应的数据为空！");
            return null;
        }

        //按行、列不同模式获取value
        Object value = null;
        if(rank == null){
            Log.e("GDataSet","rank的数据为空！");
            return null;
        }
        //行
        int rlength = 30;
        if(rank.equals(Constant.INFO_RANK_ROW)){
            List<Object> row = data.get(rIndex);
            value =  row.get(cIndex);

            rlength = row.size();
        }
        //列
        int clength = 30;
        if(rank.equals(Constant.INFO_RANK_COL)){
            List<Object> col = data.get(cIndex);
            value = col.get(rIndex);

            clength = col.size();
        }

        //获取fieldType
        List<List<String>> field = table.getField();
        List<String> colField = field.get(cIndex);
        String fieldType = colField.get(1);

        //获取flags
        Object flags = null ;
        if(table.getFlags() != null){
            String fieldName = colField.get(0);
            if(table.getFlags().get(fieldName.toLowerCase()) != null){
                List<Object> flag = (List<Object>)table.getFlags().get(fieldName.toLowerCase());
                flags = flag.get(rIndex);
            }
        }

        if(value != null){
            //cellData 按列式存储 table.cellData[cIndex][rIndex] = cData;
            if(table.getCellData() == null){
/*                GtsArrayList<GtsArrayList<TableCellData>> cellData = new
                        GtsArrayList<GtsArrayList<TableCellData>>();

                table.setCellData(cellData);*/

                table.setCellData(new TableCellData[clength][rlength]);
            }

            TableCellData[][] cellData = table.getCellData();

/*            if(cellData[cIndex] == null){
                cellData.set(cIndex,new GtsArrayList<TableCellData>());
            }*/
            if(cellData[cIndex] == null || cellData[cIndex][rIndex] == null ||
                    cellData[cIndex][rIndex] .getRawData() != value){

                TableCellData cData = new TableCellData();
                cData.setRawData(value);

                String strValue = GDataTool.getValueFormatString(value,fieldType,null,flags == null ? null : Integer.valueOf(flags.toString())); //格式化数据

                Integer colorFlags = GDataTool.GETGFLAG_COLOR((Integer) flags);  //0~3个 0 平，1 涨，2：跌 ，3：无效
                Integer isValid = GDataTool.GETGFLAG_ISINVALID((Integer) flags) == true ? 0 : 1;  //0 无效 1 有效       是否有效标识
                Integer decimal = GDataTool.GETGFLAG_DEC((Integer) flags);  //小数位 0~6位
                Integer upDownFlags = GDataTool.GET_UP_DOWN_FLAGS((Integer) flags);  //涨跌方向  0平， 1涨，2跌

                cData.setStrValue(strValue);
                cData.setColorFlags(colorFlags);
                cData.setIsValid(isValid);
                cData.setDecimal(decimal);
                cData.setUpDownFlags(upDownFlags);

/*                GtsArrayList<TableCellData> cellArr = cellData.get(cIndex);
                cellArr.set(rIndex,cData);
                cellData.set(cIndex,cellArr);*/

                TableCellData[] cellArr = cellData[cIndex];
                cellArr[rIndex] = cData;

                cellData[cIndex] = cellArr;
            }

            return cellData[cIndex][rIndex];
        }

        return null;
    }

    /**
     * 根据行号，列号获取字符串类型的数值（已格式化）
     * @param rowIndex
     * @param colIndex
     * @return
     */
    @RequiresApi(api = Build.VERSION_CODES.N)
    public String getCellText(int rowIndex, int colIndex){
        TableCellData cellData = getCellData(rowIndex,colIndex);
        if(cellData == null){
            return "-";
        }
        String strValue = cellData.getStrValue();
        return strValue;
    }


    /**
     * 根据行号，列名获取单元格对象
     * @param rowIndex
     * @param colName
     * @return
     */
    @RequiresApi(api = Build.VERSION_CODES.N)
    public TableCellData getCellDataByName(int rowIndex, String colName){
        GColumnInfo columnInfo = getColumnByName(colName);
        int colIndex = columnInfo.index;

        TableCellData cellData = getCellData(rowIndex,colIndex);
        return  cellData;
    }

    /**
     * 根据行号，列名获取字符串类型的数值（已格式化）
     * @param rowIndex
     * @param colName
     * @return
     */
    @RequiresApi(api = Build.VERSION_CODES.N)
    public String getCellTextByName(int rowIndex, String colName){
        GColumnInfo columnInfo = getColumnByName(colName);
        int colIndex = columnInfo.index;
        String strValue = getCellText(rowIndex,colIndex);

        return strValue;
    }

    /**
     * 根据列索引，获取整列数据
     * @param colIndex 列号， 从1开始
     * @return
     */
    public List<Object> getColValues(int colIndex){
        Table table = ContentMap.SOURCE_TABLE_MAP.get(id);
        if(table==null){
            Log.e("GDataSet","ContentMap.SOURCE_TABLE_MAP对应的请求id数据为空！");
            return null;
        }
        List<List<Object>> data = table.getData();
        if(data.size()==0){
            Log.e("GDataSet","table.getData()对应的数据为空！");
            return null;
        }

        //根据行、列模式获取数据
        List<Object> result = new ArrayList<>();
        if(rank == null){
            Log.e("GDataSet","rank的数据为空！");
            return null;
        }
        //行
        if(rank.equals(Constant.INFO_RANK_ROW)){
            for(List<Object> list : data){
                result.add(list.get(colIndex - 1));
            }
        }
        //列
        if(rank.equals(Constant.INFO_RANK_COL)){
            result = data.get(colIndex -1);
        }

        return result;
    }

    /**
     * 根据列名称，获取整列数据
     * @param colName
     * @return
     */
    public List<Object> getColValuesByName(String colName){
        GColumnInfo columnInfo = getColumnByName(colName);
        int colIndex = columnInfo.index;
        List<Object> result = getColValues(colIndex);

        return result;
    }

    /**
     * 根据行索引,获取整行格式化数据
     * @param rowIndex
     * @return
     */
    @RequiresApi(api = Build.VERSION_CODES.N)
    public List<String> getRowValues(int rowIndex){
        List<String> dataArr = new ArrayList<>();
        for(int k=0; k<columnInfos.size(); k++){
            String colName = columnInfos.get(k).name;

            String value = getCellTextByName(rowIndex,colName);
            dataArr.add(value);
        }

        return dataArr;
    }
}

File: demo2/app/src/main/assets/GDataSet.js

 //  问题：是不是只表示数据类型，而不表示该字段的含义. tms 本质是 int64类型， tms又表示毫秒级时间戳。建议该类型只是基础数据类型。
function GDataType(){}
GDataType.dtBool = 0;
GDataType.dtString = 1;
GDataType.dtInt = 2;
GDataType.dtInt64 = 3;
GDataType.dtFloat = 4;
GDataType.dtDouble = 5;
GDataType.dtDateTime = 6;
GDataType.dtTimeStamp = 7;
GDataType.dtTimeStampMs = 8;
GDataType.dtDataSet = 9;
GDataType.dtArray = 10;
GDataType.dtStringArray = 11;



class GColumnInfo{
    constructor() {
        this.index = null;   //列序号    int
        this.name = null;    //列名称    string
        this.dataType = null;//列类型   GDataType
        this.data = null;    //列附加数据
    }
}

class GCellData{
    
    /**
     * 获取文本类型的字符串
     * @returns {string} 格式化后的字符串
     */
    getValueText(){}

    /**
     * 获取原始数据
     * @returns {*} 原始数据
     */
    getRawData(){}


    /**
     * 
     * [0,1,1,3]
     * 
     * ------
     *  0       是否有效标识
     * ------
     *  2       0~6位 小数位
     * ------
     *  3       0~6个 颜色标志
     * ------
     *  1       0：涨   1：跌   2：平   3：无效
     * ------
     */
    /**
     * 是否有效标志
     * @returns {int} 0 无效， 1 有效
     */
    isValid(){};

    /**
     * 获取小数位
     * @returns {int} 0～6位
     */
    getDecimal(){};

    /**
     * 获取颜色标志
     * @returns {int} 0~6个颜色标志
     */
    getColorFlags(){};

    /**
     * 获取涨跌方向
     * @returns {int} 0：涨   1：跌   2：平   3：无效
     */
    getUpDownFlags(){};


}

//GDataSet
class GDataSet{
   

    constructor() {
         //id
        //集合名、表名  string
        this.name;
        //行数        int
        this.rowCount;
        //列数        int
        this.colCount;
        //开始位置
        this.begin;
        // //条数
        // this.count;
        //总数
        this.totalRows;
        //表头
        this.columnInfos;
    }

    /**
     * 根据序号获取列表信息的对象
     * @param {int} colIndex 列号
     * @returns {GColumnInfo} 表信息
     */
    getColumn(colIndex){
        for (let key in this.columnInfos) {
             if(this.columnInfos[key].index == colIndex){
                 return this.columnInfos[key];
             }
        }
    };

    /**
     * 根据名称获取列表信息的对象
     * @param {string} colName 列字段名
     * @returns {GColumnInfo} 表信息
     */
    getColumnByName(colName){
        for (const name in this.columnInfos) {
            if(colName.toLowerCase() === name.toLowerCase()){
                return this.columnInfos[name];
            }
        }
    };

    /**
     * 根据行号，列号获取单元格对象
     * @param {int} rowIndex  行号
     * @param {int} colIndex  列号
     * @return {GCellData} cellData 单元格数据
     * 
     */
    getCellData(rowIndex,colIndex){
        //单元格
        let dsManager = GDataSourceManager.sharedInstance();
        let table = dsManager.getDataSetSourceWithId(this.id);
        let rIndex = rowIndex - 1;
        let cIndex = colIndex - 1;
        let value = table.data[cIndex][rIndex];
        let fieldType = table.field[cIndex][1];
        let flags;
        if(table.flags != null) {
            let fieldName = table.field[cIndex][0];
            if(table.flags[fieldName.toLowerCase()] != null) {
               flags = table.flags[fieldName.toLowerCase()][rIndex];
            }
        }
        if(value != null){
            if(table.cellData == null){
                table.cellData = [];
            }
            if(table.cellData[cIndex] == null){
                table.cellData[cIndex]  = [];
            }
            if(table.cellData[cIndex][rIndex] == null || table.cellData[cIndex][rIndex].getRawData() != value){
                let cData = new GCellData();
                let strValue = `${GDataTool.getValueFormatString(value,fieldType,null,flags)}`;
                let rawData = value;
                let colorFlags = GDataTool.GETGFLAG_COLOR(flags); //0~3个 0 平，1 涨，2：跌 ，3：无效
                let isValid = GDataTool.GETGFLAG_ISINVALID(flags) == true ? 0 : 1; //0 无效 1 有效       是否有效标识
                let decimal = GDataTool.GETGFLAG_DEC(flags); //小数位 0~6位
                let upDownFlags = GDataTool.GET_UP_DOWN_FLAGS(flags); //涨跌方向  0平， 1涨，2跌
                cData.getValueText = ()=>{
                    return strValue;
                }
                cData.getRawData = ()=>{
                    return rawData;
                }
                cData.isValid = ()=>{
                    return isValid;
                }
                cData.getDecimal = ()=>{
                    return decimal;
                }
                cData.getColorFlags = ()=>{
                    return colorFlags;
                }
                cData.getUpDownFlags = ()=>{
                    return upDownFlags;
                }
                table.cellData[cIndex][rIndex] = cData;
            }
            return table.cellData[cIndex][rIndex];
        }
    };

    /**
     * 根据行号，列号获取字符串类型的数值（已格式化）
     * @param {int} rowIndex  行号
     * @param {int} colIndex  列号
     * @returns {string} valueStr 格式化后的值
     */
    getCellText(rowIndex,colIndex){
        let cData = this.getCellData(rowIndex,colIndex);
        if(cData == null){
            return "-";
        }
        let strValue = cData.getValueText();
        return strValue;
    };


    /**
     * 根据行号，列名获取单元格对象
     * @param {int} rowIndex    行号
     * @param {string} colName  列字段名
     * @return {GCellData} cellData 单元格数据
     */
    getCellDataByName(rowIndex,colName){
        let colInfo = this.getColumnByName(colName);
        let colIndex = colInfo.index;
        let cData = this.getCellData(rowIndex,colIndex);
        return cData;
    };

    /**
     * 根据行号，列名获取字符串类型的数值（已格式化）
     * @param {int} rowIndex    行号
     * @param {string} colName  列字段名
     * @returns {string} valueStr 格式化后的值
     */
    getCellTextByName(rowIndex,colName){
        let colInfo = this.getColumnByName(colName);
        let colIndex = colInfo.index;
        let strValue = this.getCellText(rowIndex,colIndex);
        return strValue;
    };


    /**
     * 根据列索引，获取整列数据
     * @param {int} colIndex 列号
     * @returns {Array} colArray 整列数组
     */
    getColValues(colIndex){
        let dsManager = GDataSourceManager.sharedInstance();
        let table = dsManager.getDataSetSourceWithId(this.id);
        let data = table.data;
        return data[colIndex-1];
    };

    /**
     * 根据列名称，获取整列数据
     * @param {string} colName 列字段名
     * @returns {Array} colArray 整列数组
     */
    getColValuesByName(colName){
        let colInfo = this.getColumnByName(colName);
        let data = this.getColValues(colInfo.index);
        return data;
    };


    /**
     * 根据行索引,获取整行格式化数据
     * @param {int} rowIndex 行号
     * @returns {Array} rowArray 整行数组
     */
    getRowValues(rowIndex){
        let dataArr = [];
        for(let key in this.columnInfos){
            let name = this.columnInfos[key].name;
            let value = this.getCellTextByName(rowIndex,name);
            dataArr.push(value);
        }
        return dataArr;
    }

    
}